File: backend/services/meteo.pl
import asyncio
import datetime
import math

import aiohttp
import logging

from backend.services.onairservice import OnAirService
from configuration import Topic
from backend.tools import json_serial, MQTTClient

logger = logging.getLogger("onair.meteo")

class Meteo(OnAirService):

    def __init__(self):
        self.umk_weather_url = "https://pogoda.umk.pl/api/weather"
        self.exit = False

    async def get_umk_weather(self) -> dict:
        async with aiohttp.ClientSession() as session:
            async with session.get(self.umk_weather_url) as response:
                if response.status == 200:
                    json_resp = await response.json()
                    logger.debug(await response.text())
                    data = json_resp['data']
                    return {
                        'temperature': round(float(data['tempAir200']['value']), 1),
                        'humidity': round(float(data['airHumidity']['value']), 1),
                        'pressure': round(float(data['atmosphericPressure']['value']), 1),
                        'precipitation': round(float(data['precipitation1']['value']), 1),
                        'wind': round(float(data['windSpeed']['value']), 1),
                       'date': datetime.datetime.fromisoformat(json_resp['dateUTC']).astimezone(),
                        'create_at': datetime.datetime.now(),
                    }
                else:
                    raise Exception(f"Error fetching weather data: {response.status}")

    async def run(self) -> None:
        logger.info(f"URL: {self.umk_weather_url}")
        try:
            while not self.exit:
                try:
                    weather = await self.get_umk_weather()
                    message = json_serial(weather)
                    logger.debug(message)
                    self.mqtt.publish(Topic.OnAir.format(Topic.OnAir.Facet.activity, "meteo"), message, retain=True)

                    now = datetime.datetime.now()
                    next_minute = math.ceil((now.minute + 1) / 5) * 5
                    if next_minute >= 60:
                        next_run = (now.replace(minute=0, second=0, microsecond=0) + datetime.timedelta(hours=1))
                    else:
                        next_run = now.replace(minute=next_minute, second=0, microsecond=0)
                    seconds_to_next = (next_run - now).total_seconds()
                    await asyncio.sleep(seconds_to_next)

                except Exception as e:
                    logger.error(f"Error: {e}")
        except KeyboardInterrupt:
            logger.info("Exit...")

File: backend/services/meteo.py
import asyncio
import datetime
import logging

from backend.services.meteoproviders.icm import ICMProvider
from backend.services.meteoproviders.imgw import IMGWProvider
from backend.services.meteoproviders.openmeteo import OpenMeteoProvider
from backend.services.meteoproviders.umk import UMKProvider
from backend.services.meteoproviders.visualcrossing import VisualCrossingProvider
from backend.services.onairservice import OnAirService, noexception
from configuration import Topic, Configuration
from backend.tools import json_serial
from apscheduler.schedulers.asyncio import AsyncIOScheduler
from apscheduler.triggers.cron import CronTrigger

logger = logging.getLogger("onair.meteo")

logging.getLogger('apscheduler.executors.default').setLevel(logging.WARNING)


class Meteo(OnAirService):

    def __init__(self):
        super().__init__()
        self.exit = False
        self.providers = [
            UMKProvider(),
            OpenMeteoProvider(*Configuration.location()),
            IMGWProvider(*Configuration.location()),
            ICMProvider(*Configuration.location()),
            VisualCrossingProvider(*Configuration.location(), Configuration.MAP['visualcrossing']['api_key'])]

    @noexception(logger=logger)
    async def _meteo(self, type: str, get_meteo_callback):
        first_meteo = None
        logger.debug(f"Fetching meteo {type}")
        for name in Configuration.meteo_providers()[type]:
            message = None
            try:
                logger.debug(f"Fetching meteo {type} from provider {name}")
                provider = [p for p in self.providers if p.name == name][0]
                weather = await get_meteo_callback(provider)
                if not weather:
                    raise Exception(f"Empty response when fetching meteo {type} from provider {provider.name}")
                message = json_serial({
                    'source': name,
                    'error': None,
                    'create_at': datetime.datetime.now(),
                    'data': weather,
                    })
                if first_meteo is None:
                    first_meteo = message
            except Exception as e:
                logger.error(f"Error fetching meteo {type} from provider {provider.name}: {e}")
                message - json_serial({
                    "error": f"{e}",
                    "source": provider.name,
                    "create_at": datetime.datetime.now(),
                })
            if message:
                logger.debug(message)
                self.mqtt.publish(
                    Topic.OnAir.format(Topic.OnAir.Facet.meteo, f"{type}/{provider.name}"),
                    message, retain=True)
        if first_meteo:
            self.mqtt.publish(Topic.OnAir.format(Topic.OnAir.Facet.meteo, f"{type}"), first_meteo, retain=True)

    @noexception(logger=logger)
    async def current(self) -> None:
        async def callback(provider):
            return await provider.current()
        await self._meteo('current', callback)

    @noexception(logger=logger)
    async def past(self) -> None:
        async def callback(provider):
            return await provider.past()
        await self._meteo('past/hourly', callback)

    @noexception(logger=logger)
    async def forecast(self) -> None:
        async def callback(provider):
            return await provider.forecast()
        await self._meteo('forecast/hourly', callback)

    async def run(self) -> None:
        scheduler = AsyncIOScheduler()

        every_five_min_trigger = CronTrigger.from_crontab("*/5 * * * *")  # every 5 minutes
        scheduler.add_job(self.current, every_five_min_trigger)

        every_hour_trigger = CronTrigger.from_crontab("4 * * * *")  # every hour at minute 4
        scheduler.add_job(self.past, every_hour_trigger)
        scheduler.add_job(self.forecast, every_hour_trigger)

        # run once at startup:
        scheduler.add_job(self.current)
        scheduler.add_job(self.past)
        scheduler.add_job(self.forecast)

        scheduler.start()
        await asyncio.Event().wait()
        scheduler.shutdown()
